rpc/handlers.go

```diff
@@ -1120,3 +1123,49 @@ func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id Block
 
 	return estimates, nil
 }
+
+func (h *Handler) TraceTransaction(hash felt.Felt) (TransactionTrace, *jsonrpc.Error) {
+	_, _, blockNumber, err := h.bcReader.Receipt(&hash)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+	tx, err := h.bcReader.TransactionByHash(&hash)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	id := &BlockID{Number: blockNumber - 1}
+
+	state, closer, err := h.stateByBlockID(id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer func() {
+		if closeErr := closer(); closeErr != nil {
+			h.log.Warnw("Failed to close state in starknet_call", "err", closeErr)
+		}
+	}()
+
+	header, err := h.blockHeaderByID(id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	var classes []core.Class
+	if tx, ok := tx.(*core.DeclareTransaction); ok {
+		class, stateErr := state.Class(tx.ClassHash)
+		if stateErr != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, stateErr.Error())
+		}
+		classes = append(classes, class.Class)
+	}
```

Comment:
This wont work. You need to query classes from the head state. Because `tx.ClassHash` wont be in the state for the block with number `blockNumber - 1` since it was declared on `blockNumber`

Reply:
Agree sounds reasonable, but why does it work with current implementation:
```
curl --location --request POST 'http://localhost:6060' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "starknet_traceTransaction",
    "params": {
        "transaction_hash": "0x55ba350e706f02c5903cc62f2381d6a7cc6f88efeb9c72a230487fec5f60a50"
    }
}'
```
on commit 7de8ce6a9c8bfeafe04fa5152599a04257382eca, I logged locally `append(classes, class.Class)` logic - it seems working, maybe not the way expected 😄

Reply:
Probably it worked because class is declared at 4883 block and tx is from 4968

Reply:
>  Probably it worked because class is declared at 4883 block and tx is from 4968

Yeap, indeed. It is just redeclaring a class hash. Which is fine.